subprojects/core/src/main/java/org/gradle/execution/ExcludedTaskFilteringBuildConfigurationAction.java

```diff
@@ -38,11 +39,14 @@ public void configure(BuildExecutionContext context) {
         GradleInternal gradle = context.getGradle();
         Set<String> excludedTaskNames = gradle.getStartParameter().getExcludedTaskNames();
         if (!excludedTaskNames.isEmpty()) {
-            final Set<Spec<Task>> filters = new HashSet<Spec<Task>>();
+            Multimap<GradleInternal, Spec<Task>> filters = MultimapBuilder.linkedHashKeys().hashSetValues().build();
             for (String taskName : excludedTaskNames) {
-                filters.add(taskSelector.getFilter(taskName));
+                TaskFilter filter = taskSelector.getFilter(taskName);
+                filters.put(filter.getGradle(), filter.getFilterSpec());
+            }
+            for (GradleInternal gradleInternal : filters.keySet()) {
+                gradleInternal.getTaskGraph().useFilter(Specs.intersect(filters.get(gradleInternal)));
             }
```

Comment:
I think this looks a little weird for two reasons:
- We're pushing things down into each build. 
- We're creating separate filter lists for each build. 

Could we instead add a service to the Build Tree scope that represents the task filter and then have each DefaultTaskExecutionGraph use that service as the source of filters? Then here, we wouldn't care about any particular Gradle instance, we'd just populate the "task filter service". That would remove the need for `useFilter` on the task graph.